drivers/soundwire/bus.c

```diff
@@ -1302,6 +1302,49 @@ static int sdw_handle_dp0_interrupt(struct sdw_slave *slave, u8 *slave_status)
 			clear |= SDW_DP0_INT_BRA_FAILURE;
 		}
 
+		if (status & SDW_DP0_SDCA_CASCADE) {
+			sdca_clear = sdw_read(slave, SDW_SCP_SDCA_INT1);
+			if (slave->prop.scp_sdca_int_mask1 & sdca_clear) {
+				ret = sdw_write(slave, SDW_SCP_SDCA_INT1,
+						slave->prop.scp_sdca_int_mask1);
+				if (ret < 0) {
+					dev_err(slave->bus->dev,
+						"SDW_SCP_SDCA_INT1 write failed:%d\n", ret);
+					return ret;
+				}
+			}
+			sdca_clear = sdw_read(slave, SDW_SCP_SDCA_INT2);
+			if (slave->prop.scp_sdca_int_mask2 & sdca_clear) {
+				ret = sdw_write(slave, SDW_SCP_SDCA_INT2,
+						slave->prop.scp_sdca_int_mask2);
+				if (ret < 0) {
+					dev_err(slave->bus->dev,
+						"SDW_SCP_SDCA_INT2 write failed:%d\n", ret);
+					return ret;
+				}
+			}
+			sdca_clear = sdw_read(slave, SDW_SCP_SDCA_INT3);
+			if (slave->prop.scp_sdca_int_mask3 & sdca_clear) {
+				ret = sdw_write(slave, SDW_SCP_SDCA_INT3,
+						slave->prop.scp_sdca_int_mask3);
+				if (ret < 0) {
+					dev_err(slave->bus->dev,
+						"SDW_SCP_SDCA_INT3 write failed:%d\n", ret);
+					return ret;
+				}
+			}
+			sdca_clear = sdw_read(slave, SDW_SCP_SDCA_INT4);
+			if (slave->prop.scp_sdca_int_mask4 & sdca_clear) {
+				ret = sdw_write(slave, SDW_SCP_SDCA_INT4,
+						slave->prop.scp_sdca_int_mask4);
+				if (ret < 0) {
+					dev_err(slave->bus->dev,
+						"SDW_SCP_SDCA_INT4 write failed:%d\n", ret);
+					return ret;
+				}
+			}
+		}
+
```

Comment:
I am not sure about this.
The idea is that when there is a SDCA cascade bit set, we read all 4 registers and pass them to the codec driver (be it vendor-specific or class driver), the codec driver will have information on what interrupts to look for and pass it further to the audio function drivers.
so I guess it's not really needed to have the masks here, we might as well read all the 4 registers and pass the information by extending this:

````
		/* Update the Slave driver */
		if (slave_notify && slave->ops &&
		    slave->ops->interrupt_callback) {
			slave_intr.control_port = clear;
			memcpy(slave_intr.port, &port_status,
			       sizeof(slave_intr.port));

			slave->ops->interrupt_callback(slave, &slave_intr);
		}
````

What's not clear to me is whether we clear the interrupt *before* calling the SDCA driver, or after.

Reply:
OK, got it.
Due to SDW_DP0_SDCA_CASCADE bit in DP0, do we need to ignore this bit in the sdw_handle_dp0_interrupt function?
If we don't, this function will show "Reached MAX_RETRY on DP0 read" message.
We may add the SCP_SDCA_IntStatX status into the sdw_slave_intr_status structure and pass it to the .interrupt_callback function if the vendor-specific wants to use this info.
The SCP_SDCA_IntStatX could be cleared after the .interrupt_callback function.
If this approach is good for you, I will test it tomorrow. Thanks.

Reply:
I understood. We could handle this part later.
I close this PR now.

Reply:
@shumingfan I think the bus driver should detect the SDCA cascade and we can reuse the existing interrupt callback, only with a flag that says imp-def or cascade. But I think we should the codec driver clear the interrupts it knows about.